### ISSUE TITLE
[question] How to specify settings of Clang (GNU-like) and ClangCL (MSVC-like), respectively?

### ISSUE THREAD

**Issue by hwhsu1231** [conan-io/conan#12424]
### Problem Description

As we already know, there are basically two types of Clang compilers on Windows platform:

- `clang.exe/clang++.exe`: **GNU-like** CLI Clang.
- `clang-cl.exe`: **MSVC-like** CLI Clang.

According to the [article](https://blog.conan.io/2022/10/13/Different-flavors-Clang-compiler-Windows.html) of Conan Blog, there are two example Conan Profiles, **llvm_clang.profile** and **vs_clang.profile**:

- **llvm_clang.profile**

  <details><summary>Click to expand</summary>

  ```ini
  [settings]
  os=Windows
  arch=x86_64
  build_type=Release
  compiler=clang
  compiler.version=13
  compiler.cppstd=gnu14
  compiler.runtime=dynamic
  compiler.runtime_type=Release
  compiler.runtime_version=v143
  
  [buildenv]
  PATH=+(path)C:/ws/**LLVM/Clang**13/bin
  PATH+=(path)C:/ws/msys64/mingw64/bin
  
  [conf]
  tools.env.virtualenv:auto_use=True
  ```

  </details>

- **vs_clang.profile**

  <details><summary>Click to expand</summary>

  ```ini
  [settings]
  os=Windows
  arch=x86_64
  build_type=Release
  compiler=clang
  compiler.version=14
  compiler.cppstd=gnu14
  compiler.runtime=dynamic
  compiler.runtime_type=Release
  compiler.runtime_version=v143
  
  [conf]
  tools.env.virtualenv:auto_use=True
  tools.cmake.cmaketoolchain:generator=Visual Studio 17
  ```

  </details>

However, it seems that they are the **SAME** in terms of **[settings]** section. Does that mean these two types of Clang should use the **SAME** settigns? If not, then what settings should they specify, respectively?

### Screenshots

![image](https://user-images.githubusercontent.com/71438617/198861070-9b9cd835-8d8c-4e95-8b77-e0d74407ff9c.png)
![image](https://user-images.githubusercontent.com/71438617/198861099-2ff83b54-e8b7-4da5-a91f-29749250f704.png)

**Comment by hwhsu1231** [comment]
Should **GNU-like** Clang (`clang.exe`) of LLVM or Visual Studio use the settings like **msys2_clang.profile**, **cygwin_clang.profile**...etc because they all belong to **GNU-like** Clang?

That is, the differences between **GNU-like** Clang and **MSVC-like** ClangCL in **[settings]** section are:

- **GNU-like** Clang:

  - `compiler.libcxx`

- **MSVC-like** ClangCL:

  - `compiler.runtime`
  - `compiler.runtime_type`
  - `compiler.runtime_version`

Am I correct?

**Comment by memsharded** [comment]
I think there are some possible things to clarify in the above:

- ``clang.exe`` isn't either a GNU-like or a MSVC-like compiler, it depends on the context.
- ``clang.exe`` from the Msys environment is a GNU-like compiler linking with the GNU runtime libraries
- ``clang.exe`` from the LLVM/Clang distribution is a MSVC like compiler, using the GNU-like driver (compiler flags syntax), but linking with the MSVC runtime.
- ``clang-cl.exe`` from the LLVM/CLang distribution (or MSVC distribution), is a MSVC like compiler, using the MSVC driver (compiler flags syntax) and linking with the MSVC runtime

So the difference is not in the name of the ``.exe``, but from where that exe is installed. The ``clang.exe`` from LLVM/Clang does not use ``compiler.libcxx``, because it links with the ``compiler.runtime`` from microsoft too.

**Comment by hwhsu1231** [comment]
@memsharded 

Consider the following two Clang compilers:

- `C:\Program Files\LLVM\bin\clang++.exe`
- `C:\Program Files\LLVM\bin\clang-cl.exe`

If **UserA** compiles a library called `hello` by `clang++.exe`, then **UserB** can link this `hello` library with `clang-cl.exe`. It is because `clang++.exe` and `clang-cl.exe` use the **SAME** C++ stdlib and runtime dependencies:

- MSVCP140(D).dll
- VCRUNTIME(D).dll
- ucrtbased.dll (only debug)
- api-ms-win-crt-*.dll

Therefore, we should use the **SAME** `[settings]` section for the above Clang compilers:

```ini
[settings]
os=Windows
arch=x86_64
build_type=Release
compiler=clang
compiler.version=13
compiler.cppstd=gnu14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.runtime_version=v143
```

The only **DIFFERENCE** between these two compilers is their **FRONTEND**.

- `clang++.exe`: Use GNU-like CLI (compiler flags syntax)
- `clang-cl.exe` Use MSVC-like CLI (compiler flags syntax)

Am I correct?

**Comment by memsharded** [comment]
> The only DIFFERENCE between these two compilers is their FRONTEND.

Yes, this was my understanding. However, after the blog post I was pointed out by LLVM expert that there might still be small differences internally, and it might be worth to consistently use them, and specifically to use the ``clang-cl.exe`` frontend if it is going to be mixed with msvc compiler libraries, because the fronted might still cause some small differences in the binaries.

**Comment by sykhro** [comment]
We have been using `clang-cl` with a toolchain file. The toolchain file sets up all sorts of flags and sets include paths to the `VC` runtime, allowing us to reuse it on Linux & macOS by using native `clang-cl` builds.
Our profile looks like this:

```
[settings]
os=Windows
os_build=Windows
arch=x86_64
compiler=clang
compiler.version=15

[...]

[conf]
tools.cmake.cmaketoolchain:generator=Ninja
tools.cmake.cmaketoolchain:user_toolchain = {{ [ "path/to/toolchain.cmake") ] }}
```
Is this a supported way of doing it? Shouldn't `clang-cl` pretend to be `msvc`?

**Comment by memsharded** [comment]
Following up on this old ticket.

Since then, Conan has defined the ``tools.build:compiler_executables = {"c": "clang-cl", ...`` or ``{"c": "clang"}``, to differentiate between the different LLVM/Clang frontends. There are now new examples in:
- https://docs.conan.io/2/examples/tools/autotools/llvm_clang.html
- https://docs.conan.io/2/examples/tools/cmake/cmake_toolchain/llvm_clang.html
Those examples show how it works.

Closing this ticket as solved, but please create new tickets for any further question or issue. Thanks for your feedback!
